source/common/config/watch_map.cc

```diff
@@ -0,0 +1,171 @@
+#include "common/config/watch_map.h"
+
+namespace Envoy {
+namespace Config {
+
+Watch* WatchMap::addWatch(SubscriptionCallbacks& callbacks) {
+  auto watch = std::make_unique<Watch>(callbacks);
+  Watch* watch_ptr = watch.get();
+  wildcard_watches_.insert(watch_ptr);
+  watches_.insert(std::move(watch));
+  return watch_ptr;
+}
+
+void WatchMap::removeWatch(Watch* watch) {
+  wildcard_watches_.erase(watch); // may or may not be in there, but we want it gone.
+  watches_.erase(watch);
```

Comment:
Should we assert it's in at least one of these?

Reply:
I'd say no; removeWatch is purely for cleanup. Weird as it may be for that to happen, it doesn't mean anything is wrong with the WatchMap.